app/services/invoices/progressive_billing_service.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module Invoices
+  class ProgressiveBillingService < BaseService
+    def initialize(usage_thresholds:, lifetime_usage:, timestamp: Time.current)
+      @usage_thresholds = usage_thresholds
+      @lifetime_usage = lifetime_usage
+      @timestamp = timestamp
+
+      super
+    end
+
+    def call
+      ActiveRecord::Base.transaction do
+        create_generating_invoice
+        create_threshold_fees
+        Invoices::ComputeAmountsFromFees.call(invoice:)
+        invoice.finalized!
+      end
+
+      Utils::SegmentTrack.invoice_created(invoice)
+      SendWebhookJob.perform_later('invoice.created', invoice)
+      GeneratePdfAndNotifyJob.perform_later(invoice:, email: should_deliver_email?)
+      Integrations::Aggregator::Invoices::CreateJob.perform_later(invoice:) if invoice.should_sync_invoice?
+      Integrations::Aggregator::SalesOrders::CreateJob.perform_later(invoice:) if invoice.should_sync_sales_order?
+      Invoices::Payments::CreateService.call(invoice)
+
+      result.invoice = invoice
+      result
+    rescue ActiveRecord::RecordInvalid => e
+      result.record_validation_failure!(record: e.record)
+    rescue Sequenced::SequenceError
+      raise
+    rescue => e
+      result.fail_with_error!(e)
+    end
+
+    private
+
+    attr_accessor :usage_thresholds, :lifetime_usage, :timestamp, :invoice
+
+    delegate :subscription, to: :lifetime_usage
+
+    def create_generating_invoice
+      invoice_result = CreateGeneratingService.call(
+        customer: subscription.customer,
+        invoice_type: :progressive_billing,
+        currency: usage_thresholds.first.plan.amount_currency,
+        datetime: Time.zone.at(timestamp)
+      ) do |invoice|
+        CreateInvoiceSubscriptionService
+          .call(invoice:, subscriptions: [subscription], timestamp:, invoicing_reason: :progressive_billing)
+          .raise_if_error!
+      end
+      invoice_result.raise_if_error!
+
+      @invoice = invoice_result.invoice
+    end
+
+    def create_threshold_fees
+      usage_thresholds.each do |usage_threshold|
```

Comment:
nit: would it be more clear to the reader to explicitly sort them again? (somewhere in this service, not specifically here)